feat: paste tags when pasting xblocks with tag data [FC-0049]

rpenido opened the pull request:
## Description
This PR changes the paste XBlock feature to apply tags to the destination block.

## More information
- Closes https://github.com/openedx/modular-learning/issues/180

Depends on:
- https://github.com/openedx/openedx-learning/pull/164
- https://github.com/openedx/edx-ora2/pull/2184

## Testing instructions
- Make sure the tests cover the various cases.
- Start you localdev stack on this branch
- Make sure you have Taxonomy/Tags data setup locally: https://github.com/open-craft/taxonomy-sample-data/
- Once you have a course with a bunch of units and blocks with tags applied, try copying units and various types of blocks
- Paste the blocks to another place and check if the tags are applied correctly. Note that if you paste a block from a course from one org to another, only the tags that are associated with the destination org will be applied.

___
Private ref: [FAL-3621](https://tasks.opencraft.com/browse/FAL-3621)

openedx-webhooks commented:
Thanks for the pull request, @rpenido! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

rpenido committed:
feat: paste tags when pasting xblocks with tag data

rpenido committed:
Merge branch 'master' into rpenido/fal-3621-paste-tags-when-pasting-xblocks-with-tag-data

rpenido committed:
fix: pylint

rpenido committed:
chore: temp requirements

rpenido committed:
fix: pylint

rpenido committed:
test: fix objecttag str

yusuf-musleh commented on cms/djangoapps/contentstore/views/tests/test_clipboard_paste.py:
```diff
@@ -141,6 +143,76 @@ def test_copy_and_paste_component(self, block_args):
+        # Add tags to the unit
+        taxonomy_all_org = tagging_api.create_taxonomy("test_taxonomy", "Test Taxonomy")
+
+        taxonomy_all_org = tagging_api.create_taxonomy("test_taxonomy", "Test Taxonomy")
```
This seems to be a repeated line.

rpenido replied:
Fixed: 6f76a82fafc5ea459870ca607d1f701bbcdddf90

yusuf-musleh commented on cms/djangoapps/contentstore/views/tests/test_clipboard_paste.py:
```diff
@@ -141,6 +143,76 @@ def test_copy_and_paste_component(self, block_args):
+        Tag.objects.create(taxonomy=taxonomy_all_org, value="tag_1")
+        Tag.objects.create(taxonomy=taxonomy_all_org, value="tag_2")
+        # Removing a tag is causing tag_object to fail
+        # tag_removed = Tag.objects.create(taxonomy=taxonomy, value="tag_removed")
```
Some commented out code

rpenido replied:
Removed: 6f76a82fafc5ea459870ca607d1f701bbcdddf90

yusuf-musleh commented on cms/djangoapps/contentstore/views/tests/test_clipboard_paste.py:
```diff
@@ -141,6 +143,76 @@ def test_copy_and_paste_component(self, block_args):
+        tagging_api.tag_object(
+            object_id=str(unit_key),
+            taxonomy=taxonomy_all_org,
+            # tags=["tag_1", "tag_2", "tag_removed"],
```
Also here

rpenido replied:
Removed: 6f76a82fafc5ea459870ca607d1f701bbcdddf90

yusuf-musleh reviewed (commented):
@rpenido Great work on this! So far its looking good. However I noticed when I copy a unit that contains an openassessment (open response) block and try to paste it, it fails with errors. (even when there are no tags applied.) Here's what I see in the logs:

When pasting without tags applied:

```sh
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py", line 535, in _create_block
    created_xblock, notices = import_staged_content_from_user_clipboard(
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/helpers.py", line 251, in import_staged_content_from_user_clipboard
    new_xblock = _import_xml_node_to_parent(
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/helpers.py", line 347, in _import_xml_node_to_parent
    _import_xml_node_to_parent(child_node, new_xblock, store, user_id=user_id)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/helpers.py", line 354, in _import_xml_node_to_parent
    new_xblock.add_tags_from_xml()
  File "/edx/app/edxapp/edx-platform/cms/lib/xblock/tagging/tagged_block_mixin.py", line 67, in add_tags_from_xml
    tag_data = self.xml_attributes.get('tags-v1', None) if self.xml_attributes else None
AttributeError: 'OpenAssessmentBlockWithMixins' object has no attribute 'xml_attributes'
```

When pasting with tags:
```sh
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py", line 535, in _create_block
    created_xblock, notices = import_staged_content_from_user_clipboard(
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/helpers.py", line 251, in import_staged_content_from_user_clipboard
    new_xblock = _import_xml_node_to_parent(
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/helpers.py", line 347, in _import_xml_node_to_parent
    _import_xml_node_to_parent(child_node, new_xblock, store, user_id=user_id)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/helpers.py", line 344, in _import_xml_node_to_parent
    new_xblock.sync_from_library(upgrade_to_latest=False)
  File "/edx/app/edxapp/edx-platform/xmodule/library_content_block.py", line 557, in sync_from_library
    self.get_tools(to_read_library_content=True).trigger_library_sync(
  File "/edx/app/edxapp/edx-platform/xmodule/library_tools.py", line 127, in trigger_library_sync
    if not library_api.get_v1_or_v2_library(library_key, version=library_version):
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content_libraries/api.py", line 1010, in get_v1_or_v2_library
    library = get_library(library_key)
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content_libraries/api.py", line 331, in get_library
    num_blocks = publishing_api.get_all_drafts(learning_package.id).count()
AttributeError: 'NoneType' object has no attribute 'id'
```
**edit:** One thing I forgot to mention, when the unit is pasted, the tags number+iconbutton for all the units in the subsection disappear, it's might a separate issue not related to this PR, but thought I'd mention it in case we need to create a separate ticket for it.

rpenido committed:
refactor: parse data from xml

rpenido committed:
fix: changes from review

rpenido commented:
> When pasting without tags applied:

Fixed that here: https://github.com/openedx/edx-ora2/pull/2181
Could you take a look? I'm having a problem with the tests and would appreciate an opinion!

> When pasting with tags:

I'm not able to reproduce this. Could you confirm that it is still throwing?

yusuf-musleh commented:
@rpenido 

> Could you take a look? I'm having a problem with the tests and would appreciate an opinion!

Sure thing, I left a comment regarding the tests, hopefully it helps!

> I'm not able to reproduce this. Could you confirm that it is still throwing?

So I tested with new openassessment blocks, both with and without tags, and can confirm they are working correctly. It seems like I have some openassessment blocks that are in a weird state, even on master, I am not able to view the unit that contains them, so I do not think that it's related to this PR.

yusuf-musleh reviewed (approved):
@rpenido Looks good to me! Thanks for addressing the issues. :+1:

- [X] I tested this: I confirmed pasting units/blocks with tags the tags are carried over
- [X] I read through the code
- [ ] ~I checked for accessibility issues~
- [ ] ~Includes documentation~
- [ ] ~I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.~

pomegranited commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -345,6 +345,14 @@ def _import_xml_node_to_parent(
+    from cms.lib.xblock.tagging.tagged_block_mixin import TaggedBlockMixin
+    if isinstance(new_xblock, TaggedBlockMixin):
+        new_xblock.add_tags_from_xml()
+
```
@rpenido I tested copy/pasting a tagged blocks of various types, but the pasted blocks didn't get tagged, even though I was pasting them into the same course. CC @yusuf-musleh 

These are the types I've tried:

* ORA2 works with your change from https://github.com/openedx/edx-ora2/pull/2181 ✅ 
* Text/Html ❌ 
* Zooming Image Tool (listed under Text)  ❌ 
* Video ❌ 
* Checkbox and Numerical problems (so probably all CAPA problems are affected) ❌ 
* Drag and Drop v2 ❌

rpenido replied:
Hi @pomegranited! I tested these block types with mixed results. Maybe for some cases, it's missing a refresh after paste? If that's not the case, could you help me identify the issue? Do you have any errors in the shell?

@yusuf-musleh If you have some time, could you also test to see if I have something different in my stack?

>     * Text/Html ❌

[tag_text.webm](https://github.com/openedx/edx-platform/assets/849463/d5e1ec01-e511-4125-bfa9-dfa0d7b7a8dd)
 
>     * Zooming Image Tool (listed under Text)  ❌

[tag_zooming.webm](https://github.com/openedx/edx-platform/assets/849463/9ed41551-8c49-4999-9eba-3bb56ece0028)

>     * Video ❌

[tag_video.webm](https://github.com/openedx/edx-platform/assets/849463/ecdbff3a-77b9-4e4f-893c-ef54bdd4be44)

>     * Checkbox and Numerical problems (so probably all CAPA problems are affected) ❌

The checkbox doesn't show the tags icon in the outline. The video was cut, but the copy/paste seems to work.

[tag_checkbox.webm](https://github.com/openedx/edx-platform/assets/849463/58869f52-6043-4c5a-8095-4db494d3add7)
 
>     * Drag and Drop v2 ❌

I'm unsure if it is the v2, but the Drag and Drop I tested didn't copy the tags around. I'm looking into it and will edit here if I find something.
edit: it seems to be the same case of OpenAssessments. I will implement the copy/paste the same way we did.

Is there some other XBlock that we should look for?

rpenido replied:
DragAndDrop fixed here: https://github.com/openedx/xblock-drag-and-drop-v2/pull/385

pomegranited replied:
@rpenido I don't know what was going on with my devstack yesterday, but I'm testing your branch again, and the issues I had yesterday aren't happening now. My apologies!  

> Is there some other XBlock that we should look for?

There's [a lot of XBlocks](https://openedx.atlassian.net/wiki/spaces/COMM/pages/43385346/XBlocks+Directory) that are variously maintained by Open edX or community members.. We can't possibly modify them all, and shouldn't have to either.. so I don't think we're headed down the correct path with https://github.com/openedx/xblock-drag-and-drop-v2/pull/385 and https://github.com/openedx/edx-ora2/pull/2181.

I've raised https://github.com/open-craft/edx-platform/pull/640 to handle this issue better than we've been doing it. The issue is that TaggedBlockMixin is relying on the `xml_attributes` field from XmlMixin in order to do its thing, but not all XBlocks will be XmlMixins (since `XmlMixin` is not in [XBLOCK_MIXINS](https://github.com/openedx/edx-platform/blob/ea55eca47e81bd205b423373304c9437ef95915f/cms/envs/common.py#L973-L980)), and so we can't rely on this field being present.

Instead, I've added a `tags_v1` field to TaggedBlockMixin, and use that to store the serialized tag data. What do you think?

CC @yusuf-musleh @bradenmacdonald

yusuf-musleh replied:
@pomegranited @rpenido 

Oh I see what's going on. So on `master` we have a [method](https://github.com/openedx/edx-platform/blob/master/cms/lib/xblock/tagging/tagged_block_mixin.py#L52-L57) in `TaggedBlockMixin` that would handle storing the tags in the xml for all  the "normal" blocks, since they all call the `add_xml_to_node` method, it overrides it and extends it. The HTML serialization was separate so we called the `add_tags_to_node` in there.

Seems like in this PR, we removed the `add_xml_to_node` method override from `TaggedBlockMixin`, hence we started seeing some inconsistent behavior, including blocks that do not have that method, because they would inherit it from `TaggedBlockMixin`. 

But I like @pomegranited's idea of having a new field that is separate from all that, and is more explicit on what it is and does, so it doesn't get accidentally removed cause some side effects.

yusuf-musleh replied:
@pomegranited If the tags are stored in the field rather than the xml, how will it look in terms of importing/exporting? Will the tag data still be in the olx? The reason I ask is because I remember having a [discussion](https://tasks.opencraft.com/browse/FAL-3614?focusedCommentId=284882&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284882) with @bradenmacdonald about this, and in it we discussed how users may want to be able add/modify tag data manually in their course olx before importing for example.

I just wanted to check if this also covers the case for importing/exporting.

pomegranited replied:
I didn't test the course export/import case, but the tag data _should_ still be in the OLX same as it was before.
@rpenido  could you verify this?

rpenido replied:
The OLX seems to be the same as before:
![image](https://github.com/openedx/edx-platform/assets/849463/4e82fe7a-f945-4a29-908e-ee037a3f4e11)

Thank you @pomegranited! You nailed it! I will create the PR reverting the changes in ORA, and I think we are good.

rpenido replied:
ORA revert PR here: https://github.com/openedx/edx-ora2/pull/2184

bradenmacdonald replied:
> Seems like in this PR, we removed the `add_xml_to_node` method override from `TaggedBlockMixin`

Why did we do that btw? Wasn't that working and (together with `XmlMixin`) it should cover almost all XBlocks? I don't think many XBlocks override the XML-related methods. The changes to drag-and-drop-v2 for example should definitely not have been necessary.

rpenido replied:
I think the problem is that `XmlMixin` was not in [XBLOCK_MIXINS](https://github.com/openedx/edx-platform/blob/ea55eca47e81bd205b423373304c9437ef95915f/cms/envs/common.py#L973-L980), like @pomegranited mentioned. The way it is implemented now, it doesn't rely on the `XmlMixin` and doesn't need the changes in the specific XBlock package. Do you see a better way to do it?

bradenmacdonald replied:
My concern with adding the tags as a field though, is now the tag data is going to be stored in modulestore/mongodb. So suddenly we have two sources of truth. And they could easily get out of sync. How do we know when exporting that it's using the latest tags (from the `content_tagging` API) and not the cached field data (from modulestore) from whenever the field was last updated?

I started thinking maybe we could use a `FieldOverrideProvider` or modify `CmsFieldData` to intercept this one field, and read its value from the tagging API rather than modulestore, but that would add a lot of overhead every time an XBlock is initialized or used in any way - but we only need this field during import/export.

So I'm leaning toward sticking with adding the tag field during `add_xml_to_node` and parsing it during `parse_xml`, without actually storing it as a field. I think that would work well for modulestore courses, since XML is only used during import/export anyways.

But now I realize that OLX/XML will also be used _all the time_ with Learning Core - not just for import/export but also for how the block data itself is "normally" stored. And I do feel that we probably don't want the tag data in the XML in that case (do we?) because again of the problem of two sources of truth and the need to keep it in sync. (Or do we want to allow people to change tags by editing the OLX and pressing save?)

Personally, I think the ideal would be to modify the platform so there is an extension point called for each block/node during import/export (only), to modify the OLX then.

@pomegranited @ormsbee @kdmccormick any thoughts on this? I'm feeling a bit uncertain here. It relates to a bunch of other stuff we've been discussing lately.

pomegranited replied:
@bradenmacdonald @rpenido 

> So I'm leaning toward sticking with adding the tag field during add_xml_to_node and parsing it during parse_xml, without actually storing it as a field.

Like this maybe? https://github.com/open-craft/edx-platform/pull/642/files#diff-9c6f63b94feb290ffbc18d01275713d04abc07e778fff0c42bc2a966a5c7dd6d

That PR also [adds a `studio_post_paste` handler](https://github.com/open-craft/edx-platform/pull/642/files#diff-4deb51bb5c3961a4a76ed1a264942c99df1cc63749fc6a614f8774ffe9266bcaR342) which works like the [`studio_post_duplicate` handler](https://github.com/search?q=repo%3Aopenedx%2Fedx-platform+studio_post_duplicate&type=code), so we don't have to check XBlocks to see if they're TaggedBlockMixins everywhere.

bradenmacdonald replied:
@pomegranited Yes, that looks good to me :)

bradenmacdonald replied:
Ongoing discussion on Slack too: https://openedx.slack.com/archives/C06JG6RFE4V/p1709667803595209

ormsbee replied:
@bradenmacdonald: I agree with your concerns about having multiple sources of truth. Given that this is one of those decisions that's likely to have long term repercussions, I want to make sure that we're all on the same page about why Learning Core is the way it is, and why that doesn't fit together well with the way that tagging is meant to work.

### Why store OLX?

XML is a little slow to parse and not particularly compact, so why should we store it instead of storing a more optimized format? Learning Core was designed in response to a few different problems we had with ModuleStore:

#### Exports should not need to rely on plugin code, e.g. an individual XBlock.

Content outlives software. When an XBlock is deprecated and removed, there is no way to know for certain what the serialization of that content data is supposed to look like. While most XBlocks just use the built-in defaults of having the fields expressed as attributes, there are XBlocks like ORA2 that give a much more complex and nested serialization.

This has caused issues in multiple cases where course export broke altogether for certain courses when these XBlocks were removed. The compromise behavior we have is that we now just skip trying to export block types we don't understand, but that's not a great solution, and we've seen multiple regressions in this behavior in the past.

#### Exports should be fast.

Invoking individual XBlocks at runtime to do serialization work can be slow and unpredictable. Invoking XBlock machinery also makes it memory intensive. Getting a simple dump of OLX from the database and writing it out to files will be much faster and require far less memory.

#### Exports should be robust.

When there is weird course content behavior on production, the first thing developers do is try to export a copy so that they can replicate the problem locally. Unfortunately, because the export process today requires executing a lot of XBlock plugin and runtime code, the error that's causing the erroneous behavior sometimes prevents the course from exporting altogether. At which point, trying to debug something is an infuriating exercise of trawling through the MongoDB shell on a read-replica of the prod server.

### How does Learning Core think about Truth?

As far as Learning Core is concerned right now, the true state of content are the immutable `ComponentVersions`, the `TextContents`/`FileContents` that they point to, and the `Draft` and `Published` information in the publishing app. So that's OLX and eventually other associated static assets.

Applications are encouraged to *extend* that representation with more optimized data models that are derivable from that raw OLX. So for instance, some XBlocks grab data from multiple "files", and it would be reasonable to make an `XBlockComponentVersion` model that was 1:1 with `ComponentVersion` and combined data from multiple `TextContents`. That might store the fields as JSON or something.

Another example I often use is that we might have an `ImageContent` that extends `FileContent` via a `OneToOneField` mapping, and gives metadata like width and height. We could also make these sorts of models for individual XBlocks, surfacing things like response types used, or video runtime.

These scenarios are like a progressive enhancement–they're not really adding new data to the system, but they're extracting the existing data in way that is far more useful. And we can get away with this safely because the raw data that they're enhancing stays immutable.

Whenever there's a new version, that OLX state gets written first, and everything is derived from it. We'll probably want to do that derivation synchronously, in the same transaction that we create the OLX. That way, we don't have to worry about things falling out of sync in case there's an error–the whole version creation gets rolled back.

### Why is Tagging different?

Everything I wrote above is about content itself. But tagging is living in a different lifecycle than content editing and publishing, which is where this all starts to fall apart.

As @bradenmacdonald  states in the Slack thread:

> tags live separately from the content publish history

Authors can add and remove a tag without ever publishing the component itself. Those tags immediately apply to every version of the component, whether it's been published or is still in draft form. The tags would not cause any change to the "last edited" or "last published" metadata associated with the component.

That's at the heart of Braden's point about this being a mess if it's stored in the OLX directly in Learning Core. Because if editing a tag means changing the OLX, then in Learning Core terms, that would require a new `ComponentVersion` pointing to OLX in a new `TextContent`. But tagging wants to change tagging data without ever touching those data models.

I would honestly much prefer that tagging did create new versions, so that tags are treated as peers of other attributes in the component OLX. It would be forcing tag annotations to live in the same lifecycle as content, which would simplify things like import/export. We'd have to have some code in edx-platform that was tagging aware enough to both save the appropriate attributes into the OLX and then write the matching data to the tagging API in the same transaction.

### So what now?

I don't have any great ideas. I would rather take the tags out of the attributes for individual component blocks, rather than to make it so that we have to dynamically filter them out and add them back in during export/import. Though I realize that introduces other problems around things like copy & paste.

I also worry about how tagging's decoupling from publishing means that the same library can give different results for queries based on tags, without any publishes/changes being recorded to anything inside of it. It's not enough to say, "did this collection change since the last time I checked?".

I'm also just apprehensive in ways that I can't properly articulate. One of the big headaches that we had around edx-val was trying to bridge import/export around things that were living very different lifecycles. I know that there are things that we can to do write it to a different store, but all of it just seems unnecessarily complex, and the end product is something that has more edge cases than if we just thought of tags as simple attributes with smarter post-processing.

I get that there are other types of tagging that we'll do eventually that don't live in the same lifecycle because they're made by people who don't have the same permissions–e.g. I make a personal tag for content I can't write to. But that kind of tagging wouldn't import and export with the content itself, as this does. The kind of tagging we're talking about now is editing done by the library author, that should respect import/export, and feels much cleaner to treat it more like the other attributes.

bradenmacdonald replied:
@ormsbee Unfortunately we have already gone pretty far down this path of versioning tags separately and taking an "MVP" approach, which makes any major changes challenging. For example:

* Despite all the niceties of Learning Core's new API, we need to support tagging of (modulestore) course content today, and the idea of finding a way to make tag data fit into modulestore's versioning/publishing is highly unappealing.
* Taxonomies are themselves un-versioned. So even if a tag value is stored in the OLX of a particular ComponentVersion, if you then delete the taxonomy, you're left with a broken ComponentVersion that refers to a taxonomy tag that no longer exists. This isn't the end of the world, as the OLX likely includes either the taxonomy ID or name as well as the tag value, but it would be missing key data like the parent tags of that tag. If we want to bring tags into the publishing/versioning fold, I think we may end up having to do the same thing for taxonomies. (Which _could_ be nice long-term but introduces a lot of complexity at odds with our "MVP" approach, and creates challenges around import-export.)
* We've punted on the problem of global IDs for taxonomies with the concept of an "Export ID" - the "Export ID" of a taxnonomy is used only during import/export and it allows you to do cross-instance tag sharing in a way that's robust and easy to reason about. But to facilitate that, we allow changing the Export ID at any time, and we explicitly avoid using Export IDs for anything other than import/export. So the only stable identifier for a taxonomy suitable for use in OLX _within_ a given instance its its primary key (integer) which is not suitable for usage in _cross-instance_ OLX. Of course we could change the ID architecture if needed, but again - it's another layer of wrinkles.


I do have a couple ideas that may inspire further thoughts:
1. We could get rid of the idea of including tags in the OLX. Rather, when you export a course, we can export all the tags in a single `tag.json` file (or `.csv` or `.yaml` or `.xml` or whatever). (In fact, we already have a very similar [requirement](https://github.com/openedx/modular-learning/issues/172) so most of the work to support this is already done.) For copy-paste, it's easy to deal with tags separately from the OLX. However, I think this format is less friendly to power users who are editing the OLX. (That said, so is Studio's default export format in general)
2. We could stick with the current tagging API implementation but extend it so that ObjectTags (the link between tags and content) can optionally be applied to a range of ComponentVersion objects rather than just a Component. This still has some of the problems mentioned above with un-versioned taxonomies, but it keeps the tagging API generic, while still allowing tags to be partially versioned/published along with content in the way you're describing. Partially.

ormsbee replied:
I think that I'm casting my vote for (1) then: making it so that it's not embedded in the OLX. At the end of the day, the biggest thing that worries me is the mixing and magic of making something appear as part of the versioned content (OLX) when it's really metadata over the unversioned content, and the complexity that we'll face holding that facade together over time. Doing the import/export of that data separately may not be as pretty for power users, but I think it'll really help keep things simpler in the long term.

ormsbee replied:
One day, far, far, down the line, if we want to make a Taxonomy a PublishableEntity... 😉

bradenmacdonald replied:
I'm leaning the same way now. Avoid the mismatch by not encoding tags in the OLX at all. (And consider long-term options for leveraging PublishableEntity to make everything gel together in versioned, published consistency - but a probable prerequisite or corequisite for that is using Learning Core in courses.)

ormsbee replied:
I'm curious how copy-and-paste is going to encode it though, if it's not in the OLX. Is there going to be some serialization of the tags that lives in staged content, kind of like how static assets are copied over?

Edit: To be clear, I'm definitely in favor of having tag import/export separately and not embedding into OLX. Just wondering about the details.

bradenmacdonald replied:
I was at first thinking of using an "on paste" event handler to just copy the tags over during paste. But I realize that doesn't support the "copy, delete, paste" workflow. So I think yeah, we'd add a new field to [`StagedContent`](https://github.com/openedx/edx-platform/blob/8a2c3ff14597e80be9440b8fe3d9cc8ca28cf451/openedx/core/djangoapps/content_staging/models.py#L29). It _could_ be a `tags` field but it's probably better to be a `extra_data` JSON field and then allow the `content_tagging` app to use that field to store/retrieve tags during copy/paste, so that the content_staging app doesn't need to be aware of tagging. Another option is to create a `StagedContentTags` model inside the `content_staging` app and use an "on copy" event handler to create the `StagedContentTags` model, and an "on paste" event handler to set the tags on paste.

bradenmacdonald replied:
Actually, yet another idea requires no new models or DB fields at all. During a copy event, `ObjectTag` instances could be created that point to the `StagedContent` instance, rather than pointing to a `UsageKey`. I like this approach because the tags don't have to be serialized (they are represented in the DB the same way as other "normal" tags on content). The only downside is if the `StagedContent` contains child blocks, some arbitrary new ID format is required to indicate that tags apply to child blocks inside the StagedContent.

Edit: but if we moved `StagedContent` to use `Component`s from LearningCore 😂, then we wouldn't have that problem, because each staged component would have its own DB model.

pomegranited replied:
@bradenmacdonald are you ok with the approach implemented here (for now)?

pomegranited replied:
Nudge @bradenmacdonald :)

bradenmacdonald replied:
If this is basically done, I'm OK with merging it (because nobody is using this in prod yet), but I expect that next sprint we'll be following up with a PR to change the implementation so it doesn't modify the OLX (but keeping the tests).

pomegranited commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -345,6 +345,14 @@ def _import_xml_node_to_parent(
+
+    # ToDo: Check the better place to call it
+    # I tried to call it in the xml_block.py in the parse_xml() function,
+    # but the usage_key is not persited yet there
```
nit: could you update or remove this comment?

a) it says "ToDo", but we don't have a plan to address
b) It's not correct: by the end of `xml_block.parse_xml`, the returned xblock has a location / usage key applied.

rpenido replied:
> b) It's not correct: by the end of xml_block.parse_xml, the returned xblock has a location / usage key applied.

What I mean is that at the end of `xml_block.parse_xml`, it has a usage_key in the form of `'block-v1:SampleTaxonomyOrg1+STC1+2023_1+type@html+block@localid_128844752808688'` (with localid). From what I understood, the key is persisted later and become: `'block-v1:SampleTaxonomyOrg1+STC1+2023_1+type@vertical+block@aaf8b8eb86b54281aeeab12499d2cb0b'` (the one that we want to apply the tags to).
This comment was to discuss it in the review, but I forgot to mark it here. Sorry!
Anyway, I removed it here: 2eb02303b17725cf2f5908b6b20658e0711d6d47

pomegranited replied:
Huh interesting about the "localid".. I don't know what that's for.
But thank you for removing the todo :)

pomegranited committed:
feat: duplicate tags when duplicating a tagged XBlock. (#637)

* feat: duplicate tags when duplicating a tagged XBlock.

* test: fixes flaky test

rpenido committed:
chore: update openedx-learning version

rpenido committed:
fix: removing temp comment

rpenido committed:
test: fix tag_object param order

rpenido committed:
chore: update ora2 version

rpenido committed:
Merge branch 'master' into rpenido/fal-3621-paste-tags-when-pasting-xblocks-with-tag-data

rpenido committed:
chore: fix inconsistent python dependencies

rpenido committed:
chore: temp update xblock-drag-and-drop-v2 version

rpenido committed:
Merge branch 'master' into rpenido/fal-3621-paste-tags-when-pasting-xblocks-with-tag-data

rpenido committed:
Revert "chore: temp update xblock-drag-and-drop-v2 version"

This reverts commit dacba7d47740900d4536bd11cca856ff862dd7e7.

pomegranited committed:
fix: store tags in TaggedBlockMixin.tags_v1 field

Previously, we relied on the presence of the xml_attributes field, which
is added by the XMLMixin. However, XMLMixin is not part of the default
XBLOCK_MIXINS, so custom XBlocks weren't able to copy/paste tags.

rpenido committed:
chore: update ora2 to reverted version

rpenido committed:
style: rename add_tags_from_xml() to add_tags_from_field()

rpenido committed:
chore: temp ora2 update

rpenido committed:
Merge branch 'master' into rpenido/fal-3621-paste-tags-when-pasting-xblocks-with-tag-data

pomegranited committed:
refactor: TaggedBlockMixin (#642)

* refactor: TaggedBlockMixin overrides add_xml_to_node

so we can use inheritance to include tags in exported blocks

* refactor: store tags_v1 in an internal field

instead of an XBlock field, because we don't want this information
persisted to the modulestore.

* refactor: adds studio_post_paste handler

for use after copy/pasting content from the clipboard.

* fix: make serialize_tag_data a classmethod

so we can serialize tags for blocks that aren't officially TaggedXBlocks

pomegranited commented on openedx/core/djangoapps/content_tagging/api.py:
```diff
@@ -161,16 +162,42 @@ def get_all_object_tags(
     for object_id, block_tags in groupby(all_object_tags, lambda x: x.object_id):
         grouped_object_tags[object_id] = {}
-        for taxonomy_id, taxonomy_tags in groupby(block_tags, lambda x: x.tag.taxonomy_id):
+        for taxonomy_id, taxonomy_tags in groupby(block_tags, lambda x: x.tag.taxonomy_id if x.tag else 0):
```
The ObjectTag.objects.filter above guarantees that `tag` is not null.. why do we need to check that again here (and assert below)?

rpenido replied:
This is a type guard to make our typer checker happy. I usually use assert for type guards (without error-handling routines), as this is not expected to be thrown at all.

Is there a better way to handle this?

pomegranited replied:
I'm mainly puzzled as to why we had to add these checks now, but they weren't needed before. But it's not a big deal.

rpenido replied:
Basically, we "opt-in" to type checking when we add a return type to the function.
Accessing `x.tag.taxonomy_id` will always throw a type error if the tag is nullable and we didn't check it before calling it.

pomegranited commented on openedx/core/djangoapps/content_tagging/utils.py:
```diff
@@ -30,11 +32,10 @@ def get_content_key_from_string(key_str: str) -> ContentKey:
+def get_context_key_from_key(content_key: ContentKey) -> ContextKey:
     """
-    Get context key from an key string
+    Get context key from an key
```
nit:
```suggestion
    Returns the context key from a given content key.
```

rpenido replied:
Done: 0f2ac20831a755612562dfcdad426d6e7e3e44b7

pomegranited reviewed (approved):
I've left a question and a nit about a comment, but this is working beautifully @rpenido !

:+1:

- [x] I tested this on my devstack in Studio:
   * Created a course and a unit
   * Added blocks to the unit of type: html, video, CAPA, drag and drop, survey/poll, ORA
   * Added tags to all of the above blocks
   * Copy/pasted these tagged blocks into a new unit for the same course -- tags were copied too.
   * Tested copy/pasting untagged blocks -- still works as expected
   * Tested copy/pasting tagged units into a course with a different org -- disallowed tags were omitted on paste, and allowed tags were copied as expected.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

rpenido committed:
docs: improve docstring

Co-authored-by: Jillian <jill@opencraft.com>

rpenido committed:
chore: update edx-ora2 package

rpenido commented:
Updated ora2 package: b4f195a6406b11e3ba1369883574a61b1fbabb74 CC @pomegranited

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -1106,6 +1107,57 @@ def test_duplicate_library_content_block(self):  # pylint: disable=too-many-stat
+        # Refresh.
+        source_chapter = self.store.get_item(source_chapter.location)
+        expected_chapter_tags = 'A:one,two;B:four,three'
+        assert source_chapter.serialize_tag_data(source_chapter.location) == expected_chapter_tags
```
Can you just change this test so that it uses `tagging_api.get_object_tags` to get the tags, rather than `. serialize_tag_data()` ? 

Because `serialize_tag_data` assumes that we're using `TaggedBlockMixin` but that's an implementation detail that we may change in the near future, based on that discussion.

What I'd like to see is that even if we change to not include tags in the OLX, these tests will still be working and not need any changes.

rpenido replied:
Done: 672f76cd7d8e54c0e6814be3616882d89857cb29

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_clipboard_paste.py:
```diff
@@ -141,6 +143,69 @@ def test_copy_and_paste_component(self, block_args):
+        tags = list(tagging_api.get_object_tags(str(dest_unit_key)))
+        assert len(tags) == 2
+        assert str(tags[0]) == f'<ObjectTag> {dest_unit_key}: test_taxonomy=tag_1'
+        assert str(tags[1]) == f'<ObjectTag> {dest_unit_key}: test_taxonomy=tag_2'
```
^ This test is great, and even if we end up changing how we handle the tags when copying and pasting, this test shouldn't need to be modified.

rpenido committed:
test: change xblock duplication tests to be implementation independent

openedx-webhooks commented:
@rpenido 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34270&entry.1484655318=https://github.com/rpenido&entry.752974735=2024-02-21+20:58&entry.1917517419=2024-03-08+20:03&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
